### ISSUE TITLE
HAVE_POSIX_STRERROR_R is defined too late

### ISSUE THREAD

**Issue by JonasVautherin** [curl/curl#2582]
### I did this
Cross-compile curl using dockcross for armv6 using the following script:

```sh
#!/bin/sh

# Prepare the dockcross script (https://github.com/dockcross/dockcross):
docker run --rm dockcross/linux-armv6 > ./dockcross
chmod +x ./dockcross

# Clone curl
git clone --depth 1 https://github.com/curl/curl

# Run cmake
cd curl
../dockcross cmake -DCMAKE_USE_OPENSSL=OFF -DCURL_CA_PATH_SET=OFF -DHAVE_POSIX_STRERROR_R=1 -Bbuild -H.

# Run make
../dockcross make -Cbuild
```

### I expected the following

A successful build

### I experienced the following

```
[  6%] Building C object lib/CMakeFiles/libcurl.dir/strerror.c.o
/work/lib/strerror.c:32:6: error: #error "strerror_r MUST be either POSIX, glibc or vxworks-style"
 #    error "strerror_r MUST be either POSIX, glibc or vxworks-style"
      ^
lib/CMakeFiles/libcurl.dir/build.make:660: recipe for target 'lib/CMakeFiles/libcurl.dir/strerror.c.o' failed
make[2]: *** [lib/CMakeFiles/libcurl.dir/strerror.c.o] Error 1
make[2]: Leaving directory '/work/build'
```

### I identified the following problem

I believe `HAVE_POSIX_STRERROR_R` should be set to 1 for this platform, which is why I pass `-DHAVE_POSIX_STRERROR_R=1` [in my script](https://ptpb.pw/5umC/sh#L-12). But my argument is somehow ignored, and the build still fails. If I run `cmake` twice, as in [this script](https://ptpb.pw/DMnb/sh), then it works.

Looking into the CMakeLists.txt, I realized that commenting [this line](https://github.com/curl/curl/blob/master/CMakeLists.txt#L981) out was solving the issue. I don't know how `CURL_TEST` works, but it seems to me that `HAVE_POSIX_STRERROR_R` is not set until [line 1156](https://github.com/curl/curl/blob/master/CMakeLists.txt#L1156), as follows:

1. `HAVE_POSIX_STRERROR_R` is defined in [`curl_config.h.cmake`](https://github.com/curl/curl/blob/master/lib/curl_config.h.cmake#L486).
2. `curl_config.h.cmake` is generated from [`lib/CMakeLists.txt`](https://github.com/curl/curl/blob/master/lib/CMakeLists.txt#L3).
3. `lib/CMakeLists.txt` is called after the `CURL_TEST` in the main [`CMakeLists.txt`](https://github.com/curl/curl/blob/master/CMakeLists.txt#L1156).

### possible fixes

1. Run this [`TEST_CURL`](https://github.com/curl/curl/blob/master/CMakeLists.txt#L981) after [`add_subdirectory(lib)`](https://github.com/curl/curl/blob/master/CMakeLists.txt#L1156) (this worked for me).
2. Run `add_subdirectory(lib)` somewhere before `TEST_CURL` (I did not manage to do that)

I would be glad to write a PR if you believe one of those fixes makes sense. Note that I suspect this bug may happen for other `VARs` as well (e.g. `HAVE_GLIBC_STRERROR_R`).

### curl/libcurl version

master branch

### operating system

Arch Linux

**Comment by jzakrzewski** [comment]
I believe the right way would be to check if the variable is already defined before doing any tests. Internally `curl_internal_test_run` skips compilation of tests if they were already compiled. In this case they weren't but the variable is set and that alone should disable the reevaluation.

**Comment by JonasVautherin** [comment]
@jzakrzewski I tried to do what I understood from your comment in #2583. Is it what you meant? At least it works for me.

**Comment by snikulov** [comment]
@JonasVautherin BTW, #2443 related to cross-compilation.
Could you please check those changes before we'll integrate them?

**Comment by JonasVautherin** [comment]
@snikulov #2443 seems to fix my issue as well (I ran [my script](https://ptpb.pw/cOAG/sh) on https://github.com/MaxSavenkov/curl) without using `-DHAVE_POSIX_STRERROR_R=1`.

The script:

```sh
#!/bin/sh

# Prepare the dockcross script (https://github.com/dockcross/dockcross):
docker run --rm dockcross/linux-armv6 > ./dockcross
chmod +x ./dockcross

# Clone curl
git clone https://github.com/MaxSavenkov/curl

# Run cmake
cd curl
../dockcross cmake -DCMAKE_USE_OPENSSL=OFF -DCURL_CA_PATH_SET=OFF -Bbuild -H.

# Run make
../dockcross make -Cbuild
```

**Comment by snikulov** [comment]
Will be closed per #2443

**Comment by snikulov** [comment]
Fixed on master with 7e93637acd9f5741ac4c09bbca353ac8da42bb17

**Comment by bagder** [comment]
`//` comments, really? Won't that cause problems if you want to have the compiler be picky and check for C89 issues (as that's our lowest denominator still) ?

(and please make the commit message follow our template)
